File: lsd-interceptors/src/main/java/io/lsdconsulting/interceptors/http/common/DefaultHttpInteractionHandler.java
package io.lsdconsulting.interceptors.http.common;

import com.lsd.core.IdGenerator;
import com.lsd.core.LsdContext;
import io.lsdconsulting.interceptors.common.Headers;
import io.lsdconsulting.interceptors.http.naming.DestinationNameMappings;
import io.lsdconsulting.interceptors.http.naming.SourceNameMappings;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.time.Duration;
import java.util.Map;

import static com.lsd.core.builders.ActivateLifelineBuilder.activation;
import static com.lsd.core.builders.DeactivateLifelineBuilder.deactivation;
import static com.lsd.core.builders.MessageBuilder.messageBuilder;
import static com.lsd.core.domain.MessageType.SYNCHRONOUS;
import static com.lsd.core.domain.MessageType.SYNCHRONOUS_RESPONSE;
import static j2html.TagCreator.*;
import static java.lang.System.lineSeparator;
import static java.util.stream.Collectors.joining;
import static lsd.format.PrettyPrinterKt.prettyPrint;
import static org.apache.commons.lang3.StringUtils.isEmpty;

@ToString
@EqualsAndHashCode
public class DefaultHttpInteractionHandler implements HttpInteractionHandler {

    private final LsdContext lsdContext;
    private final SourceNameMappings sourceNameMappings;
    private final DestinationNameMappings destinationNameMappings;
    private final IdGenerator idGenerator;

    public DefaultHttpInteractionHandler(LsdContext lsdContext, SourceNameMappings sourceNameMappings, DestinationNameMappings destinationNameMappings) {
        this.lsdContext = lsdContext;
        this.sourceNameMappings = sourceNameMappings;
        this.destinationNameMappings = destinationNameMappings;
        this.idGenerator = lsdContext.getIdGenerator();
    }

    @Override
    public void handleRequest(String method, Map<String, String> requestHeaders, String path, String body) {
        var targetName = deriveTargetName(requestHeaders, path);
        lsdContext.capture(
                messageBuilder()
                        .id(idGenerator.next())
                        .from(deriveSourceName(requestHeaders, path))
                        .to(targetName)
                        .label(method + " " + path)
                        .data(renderHtmlFor(path, requestHeaders, null, prettyPrint(body), null))
                        .type(SYNCHRONOUS)
                        .build(),
                activation().of(targetName).colour("skyblue").build()
        );
    }

    @Override
    public void handleResponse(String statusMessage, Map<String, String> requestHeaders, Map<String, String> responseHeaders, String path, String body, Duration duration) {
        String colour = "";
        if (statusMessage.startsWith("4") || statusMessage.startsWith("5")) colour = "red";

        var targetName = deriveTargetName(requestHeaders, path);
        lsdContext.capture(
                messageBuilder()
                        .id(idGenerator.next())
                        .from(targetName)
                        .to(deriveSourceName(requestHeaders, path))
                        .label(statusMessage + " (" + duration.toMillis() + "ms)")
                        .data(renderHtmlFor(path, requestHeaders, responseHeaders, prettyPrint(body), duration))
                        .type(SYNCHRONOUS_RESPONSE)
                        .colour(colour)
                        .duration(duration)
                        .build(),
                deactivation().of(targetName).build()
        );
    }

    private String renderHtmlFor(String path, Map<String, String> requestHeaders, Map<String, String> responseHeaders, String prettyBody, Duration duration) {
        return div(
                section(
                        h3("Full Path"),
                        span(path)
                ), isMissingHeaders(requestHeaders)
                        ? p()
                        : section(h3("Request Headers"), p(prettyPrintHeaders(requestHeaders)))
                , isMissingHeaders(responseHeaders)
                        ? p()
                        : section(h3("Response Headers"), p(prettyPrintHeaders(responseHeaders)))
                , isEmpty(prettyBody)
                        ? p()
                        : section(h3("Body"), p(prettyBody)
                , null == duration
                        ? p()
                        : section(h3("Duration"), p(duration.toMillis() + "ms"))
                )
        ).render();
    }

    private static boolean isMissingHeaders(Map<String, String> requestHeaders) {
        return requestHeaders == null || requestHeaders.isEmpty();
    }

    private String prettyPrintHeaders(Map<String, String> headers) {
        return headers.entrySet().stream().map(entry ->
                        entry.getKey() + ": " + entry.getValue())
                .collect(joining(lineSeparator()));
    }

    private String deriveTargetName(Map<String, String> headers, String path) {
        return headers.containsKey(Headers.HeaderKeys.TARGET_NAME.key())
                ? headers.get(Headers.HeaderKeys.TARGET_NAME.key())
                : destinationNameMappings.mapForPath(path);
    }

    private String deriveSourceName(Map<String, String> headers, String path) {
        return headers.containsKey(Headers.HeaderKeys.SOURCE_NAME.key())
                ? headers.get(Headers.HeaderKeys.SOURCE_NAME.key())
                : sourceNameMappings.mapForPath(path);
    }
}

File: lsd-interceptors/src/main/java/io/lsdconsulting/interceptors/http/common/DefaultHttpInteractionHandler.kt
package io.lsdconsulting.interceptors.http.common

import com.lsd.core.IdGenerator
import com.lsd.core.LsdContext
import com.lsd.core.builders.ActivateLifelineBuilder
import com.lsd.core.builders.DeactivateLifelineBuilder
import com.lsd.core.builders.MessageBuilder.Companion.messageBuilder
import com.lsd.core.domain.MessageType
import io.lsdconsulting.interceptors.common.Headers
import io.lsdconsulting.interceptors.common.Headers.HeaderKeys.TARGET_NAME
import io.lsdconsulting.interceptors.http.naming.DestinationNameMappings
import io.lsdconsulting.interceptors.http.naming.SourceNameMappings
import j2html.TagCreator
import lsd.format.prettyPrint
import org.apache.commons.lang3.StringUtils
import java.time.Duration
import java.util.stream.Collectors

data class DefaultHttpInteractionHandler(
    private val lsdContext: LsdContext,
    private val sourceNameMappings: SourceNameMappings,
    private val destinationNameMappings: DestinationNameMappings
) : HttpInteractionHandler {
    private val idGenerator: IdGenerator = lsdContext.idGenerator

    override fun handleRequest(method: String, requestHeaders: Map<String, String>, path: String, body: String) {
        val targetName = deriveTargetName(requestHeaders, path)
        lsdContext.capture(
            messageBuilder()
                .id(idGenerator.next())
                .from(deriveSourceName(requestHeaders, path))
                .to(targetName)
                .label("$method $path")
                .data(renderHtmlFor(path, requestHeaders, null, prettyPrint(body), null))
                .type(MessageType.SYNCHRONOUS)
                .build(),
            ActivateLifelineBuilder.activation().of(targetName).colour("skyblue").build()
        )
    }

    override fun handleResponse(
        statusMessage: String,
        requestHeaders: Map<String, String>,
        responseHeaders: Map<String, String>,
        path: String,
        body: String,
        duration: Duration
    ) {
        var colour = ""
        if (statusMessage.startsWith("4") || statusMessage.startsWith("5")) colour = "red"
        val targetName = deriveTargetName(requestHeaders, path)
        lsdContext.capture(
            messageBuilder()
                .id(idGenerator.next())
                .from(targetName)
                .to(deriveSourceName(requestHeaders, path))
                .label(statusMessage + " (" + duration.toMillis() + "ms)")
                .data(renderHtmlFor(path, requestHeaders, responseHeaders, prettyPrint(body), duration))
                .type(MessageType.SYNCHRONOUS_RESPONSE)
                .colour(colour)
                .duration(duration)
                .build(),
            DeactivateLifelineBuilder.deactivation().of(targetName).build()
        )
    }

    private fun renderHtmlFor(
        path: String,
        requestHeaders: Map<String, String>,
        responseHeaders: Map<String, String>?,
        prettyBody: String,
        duration: Duration?
    ): String {
        return TagCreator.div(
            TagCreator.section(
                TagCreator.h3("Full Path"),
                TagCreator.span(path)
            ),
            if (isMissingHeaders(requestHeaders)) TagCreator.p() else TagCreator.section(
                TagCreator.h3("Request Headers"),
                TagCreator.p(prettyPrintHeaders(requestHeaders))
            ),
            if (isMissingHeaders(responseHeaders)) TagCreator.p() else TagCreator.section(
                TagCreator.h3("Response Headers"),
                TagCreator.p(prettyPrintHeaders(responseHeaders))
            ),
            if (StringUtils.isEmpty(prettyBody)) TagCreator.p() else TagCreator.section(
                TagCreator.h3("Body"),
                TagCreator.p(prettyBody),
                if (null == duration) TagCreator.p() else TagCreator.section(
                    TagCreator.h3("Duration"),
                    TagCreator.p(duration.toMillis().toString() + "ms")
                )
            )
        ).render()
    }

    private fun prettyPrintHeaders(headers: Map<String, String>?): String {
        return headers!!.entries.stream().map { (key, value): Map.Entry<String, String> -> "$key: $value" }
            .collect(Collectors.joining(System.lineSeparator()))
    }

    private fun deriveTargetName(headers: Map<String, String>, path: String): String {
        return if (headers.containsKey(TARGET_NAME.key())) headers[TARGET_NAME.key()]!! else destinationNameMappings.mapForPath(
            path
        )
    }

    private fun deriveSourceName(headers: Map<String, String>, path: String): String {
        return if (headers.containsKey(Headers.HeaderKeys.SOURCE_NAME.key())) headers[Headers.HeaderKeys.SOURCE_NAME.key()]!! else sourceNameMappings.mapForPath(
            path
        )
    }

    companion object {
        private fun isMissingHeaders(requestHeaders: Map<String, String>?): Boolean {
            return requestHeaders.isNullOrEmpty()
        }
    }
}
